File: ios-app/BayitPlusApp/Models/SeriesModels.swift
import Foundation

// MARK: - Series

/// Response from GET /api/v1/content/series/{series_id}
struct SeriesDetail: Decodable, Sendable, Identifiable {
    let id: String
    let title: String?
    let description: String?
    let thumbnail: String?
    let backdrop: String?
    let category: String?
    let year: Int?
    let rating: String?
    let genre: String?
    let cast: [String]?
    let director: String?
    let totalSeasons: Int?
    let totalEpisodes: Int?
    let trailerUrl: String?
    let trailerStreamUrl: String?
    let previewUrl: String?
    let tmdbId: String?
    let imdbId: String?
    let availableSubtitleLanguages: [String]?
    let hasSubtitles: Bool?
    let isKidsContent: Bool?
    let ageRating: String?
    let seasons: [SeasonSummary]?
    let related: [RelatedItem]?

    private struct FlexRating: Decodable {
        let value: String?

        init(from decoder: Decoder) throws {
            let container = try decoder.singleValueContainer()
            if let str = try? container.decode(String.self) {
                value = str
            } else if let num = try? container.decode(Double.self) {
                value = String(format: "%.1f", num)
            } else {
                value = nil
            }
        }
    }

    enum CodingKeys: String, CodingKey {
        case id, title, description, thumbnail, backdrop, category, year
        case rating, genre, cast, director
        case totalSeasons, totalEpisodes
        case trailerUrl, trailerStreamUrl, previewUrl
        case tmdbId, imdbId
        case availableSubtitleLanguages, hasSubtitles
        case isKidsContent, ageRating, seasons, related
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decode(String.self, forKey: .id)
        title = try c.decodeIfPresent(String.self, forKey: .title)
        description = try c.decodeIfPresent(String.self, forKey: .description)
        thumbnail = try c.decodeIfPresent(String.self, forKey: .thumbnail)
        backdrop = try c.decodeIfPresent(String.self, forKey: .backdrop)
        category = try c.decodeIfPresent(String.self, forKey: .category)
        year = try c.decodeIfPresent(Int.self, forKey: .year)
        genre = try c.decodeIfPresent(String.self, forKey: .genre)
        cast = try c.decodeIfPresent([String].self, forKey: .cast)
        director = try c.decodeIfPresent(String.self, forKey: .director)
        totalSeasons = try c.decodeIfPresent(Int.self, forKey: .totalSeasons)
        totalEpisodes = try c.decodeIfPresent(Int.self, forKey: .totalEpisodes)
        trailerUrl = try c.decodeIfPresent(String.self, forKey: .trailerUrl)
        trailerStreamUrl = try c.decodeIfPresent(String.self, forKey: .trailerStreamUrl)
        previewUrl = try c.decodeIfPresent(String.self, forKey: .previewUrl)
        tmdbId = try c.decodeIfPresent(String.self, forKey: .tmdbId)
        imdbId = try c.decodeIfPresent(String.self, forKey: .imdbId)
        availableSubtitleLanguages = try c.decodeIfPresent([String].self, forKey: .availableSubtitleLanguages)
        hasSubtitles = try c.decodeIfPresent(Bool.self, forKey: .hasSubtitles)
        isKidsContent = try c.decodeIfPresent(Bool.self, forKey: .isKidsContent)
        ageRating = try c.decodeIfPresent(String.self, forKey: .ageRating)
        seasons = try c.decodeIfPresent([SeasonSummary].self, forKey: .seasons)
        related = try c.decodeIfPresent([RelatedItem].self, forKey: .related)
        rating = try c.decodeIfPresent(FlexRating.self, forKey: .rating)?.value
    }
}

/// Summary of a season within a series
struct SeasonSummary: Decodable, Sendable, Identifiable {
    let seasonNumber: Int
    let episodeCount: Int
    let firstEpisodeId: String?
    let firstEpisodeThumbnail: String?

    var id: Int {
        seasonNumber
    }
}

/// Response from GET /api/v1/content/series/{series_id}/season/{season_num}/episodes
struct SeasonEpisodesResponse: Decodable, Sendable {
    let seriesId: String
    let seasonNumber: Int
    let episodes: [EpisodeItem]
}

/// An episode within a season
struct EpisodeItem: Decodable, Sendable, Identifiable {
    let id: String
    let title: String?
    let description: String?
    let thumbnail: String?
    let episodeNumber: Int?
    let duration: String?
    let previewUrl: String?
    let streamUrl: String?
    let directUrl: String?
    let isTranscoded: Bool?
}

/// Response from GET /api/v1/content/series
struct SeriesListResponse: Decodable, Sendable {
    let items: [ContentItem]
    let total: Int
    let page: Int
    let limit: Int
}

File: android-app/core/core-model/src/main/java/tv/bayit/plus/core/model/SeriesModels.kt
package tv.bayit.plus.core.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/** Response from GET /api/v1/content/series/{series_id} */
@Serializable
data class SeriesDetail(
    val id: String,
    val title: String? = null,
    val description: String? = null,
    val thumbnail: String? = null,
    val backdrop: String? = null,
    val category: String? = null,
    val year: Int? = null,
    val rating: String? = null,
    val genre: String? = null,
    val cast: List<String>? = null,
    val director: String? = null,
    @SerialName("total_seasons") val totalSeasons: Int? = null,
    @SerialName("total_episodes") val totalEpisodes: Int? = null,
    @SerialName("trailer_url") val trailerUrl: String? = null,
    @SerialName("preview_url") val previewUrl: String? = null,
    @SerialName("tmdb_id") val tmdbId: String? = null,
    @SerialName("imdb_id") val imdbId: String? = null,
    @SerialName("available_subtitle_languages")
    val availableSubtitleLanguages: List<String>? = null,
    @SerialName("has_subtitles") val hasSubtitles: Boolean? = null,
    @SerialName("is_kids_content") val isKidsContent: Boolean? = null,
    @SerialName("age_rating") val ageRating: String? = null,
    val seasons: List<SeasonSummary>? = null,
    val related: List<RelatedItem>? = null,
)

/** Summary of a season within a series. */
@Serializable
data class SeasonSummary(
    @SerialName("season_number") val seasonNumber: Int,
    @SerialName("episode_count") val episodeCount: Int,
    @SerialName("first_episode_id") val firstEpisodeId: String? = null,
    @SerialName("first_episode_thumbnail") val firstEpisodeThumbnail: String? = null,
)

/** Response from GET /api/v1/content/series/{series_id}/season/{season_num}/episodes */
@Serializable
data class SeasonEpisodesResponse(
    @SerialName("series_id") val seriesId: String,
    @SerialName("season_number") val seasonNumber: Int,
    val episodes: List<EpisodeItem> = emptyList(),
)

/** An episode within a season. */
@Serializable
data class EpisodeItem(
    val id: String,
    val title: String? = null,
    val description: String? = null,
    val thumbnail: String? = null,
    @SerialName("episode_number") val episodeNumber: Int? = null,
    val duration: String? = null,
    @SerialName("preview_url") val previewUrl: String? = null,
    @SerialName("stream_url") val streamUrl: String? = null,
    @SerialName("direct_url") val directUrl: String? = null,
    @SerialName("is_transcoded") val isTranscoded: Boolean? = null,
)

/** Response from GET /api/v1/content/series */
@Serializable
data class SeriesListResponse(
    val items: List<ContentItem> = emptyList(),
    val total: Int,
    val page: Int,
    val limit: Int,
)
